src/components/lab/NumberInput/NumberInput.tsx

```diff
@@ -0,0 +1,312 @@
+    defaultValue?: number;
+    /** Fires when the input’s value is changed by the user. Provides new value as an callback's argument */
+    onUpdate?: (value: number | undefined) => void;
+}
+
+const voidFunction = (..._props: any[]) => {};
+
+// Useful when in input string '-1.' is typed and value={-1} prop passed.
+// In this case we leave input string without replacing it by '-1'.
+// Means that where is no need for replacing current input value with external value
+function areStringRepresentationOfNumbersEqual(v1: string, v2: string) {
+    if (v1 === v2) {
+        return true;
+    }
+
+    const {valid: v1Valid, value: v1Value} = getParsedValue(v1);
+    const {valid: v2Valid, value: v2Value} = getParsedValue(v2);
+
+    if (v1Valid && v2Valid) {
+        return v1Value === v2Value;
+    }
+
+    const v1OnlyNumbers = v1.replace(/\D/g, '');
+    const v2OnlyNumbers = v2.replace(/\D/g, '');
+
+    if (v1OnlyNumbers.length === v2OnlyNumbers.length && v1OnlyNumbers.length === 0) {
+        // exmpl, when just '-' typed and '' (equivalent for undefined) value passed
+        return true;
+    }
+    return false;
+}
+
+function getStringValue(value: number | undefined) {
+    return value ? String(value) : '';
+}
+
+export const NumberInput = React.forwardRef<HTMLSpanElement, NumberInputProps>(function NumberInput(
+    {endContent, ...props},
+    ref,
+) {
+    const {
+        value: externalValue,
+        defaultValue: externalDefaultValue,
+        onChange: handleChange,
+        onUpdate,
+        min: externalMin,
+        max: externalMax,
+        shiftMultiplier: externalShiftMultiplier = 10,
+        step: externalStep = 1,
+        size,
+        view,
+        disabled,
+        hasControls = true,
+        validationState,
+        onFocus,
+        onBlur,
+        onKeyDown,
+        onKeyUp,
+        allowMouseWheel = false,
+        allowDecimal = false,
+        className,
+    } = {...defaultBaseTextInputProps, ...props};
+
+    const {
+        min,
+        max,
+        step: baseStep,
+        value,
+        defaultValue,
+        shiftMultiplier,
+    } = getInternalState({
+        min: externalMin,
+        max: externalMax,
+        step: externalStep,
+        shiftMultiplier: externalShiftMultiplier,
+        allowDecimal,
+        value: externalValue,
+        defaultValue: externalDefaultValue,
+    });
+
+    const [inputValue, setInputValue] = React.useState(
+        getStringValue(value) ?? getStringValue(defaultValue),
+    );
+
+    React.useEffect(() => {
+        const stringPropsValue = getStringValue(value);
+        setInputValue((currentInputValue) => {
+            if (!areStringRepresentationOfNumbersEqual(currentInputValue, stringPropsValue)) {
+                return stringPropsValue;
+            }
+            return currentInputValue;
+        });
+    }, [value]);
+
+    const clamp = true;
+    const [step, setStep] = React.useState(baseStep);
```

Comment:
I thought that it would be helpful for determining current step while pressing arrow buttons (controls in component) with shift button pressed (realized) 
But then I realized, that it would require adding new wrapperRef prop for detecting focusWithin, and change step on focusWithin state change. 
If you think that shift pressing shouldn't change the step value for component
If I'm not mistaken, pointer events don't have info, if shift is pressed or not now